flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/VarCharType.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.logical;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.api.ValidationException;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Logical type of a variable-length character string.
+ *
+ * <p>The serialized string representation is {@code VARCHAR(n)} where {@code n} is the maximum
+ * number of code points. {@code n} must have a value between 1 and {@link Integer#MAX_VALUE} (both
+ * inclusive). If no length is specified, {@code n} is equal to 1.
+ */
+@PublicEvolving
+public final class VarCharType extends LogicalType {
+
+	private static final int MIN_LENGTH = 1;
+
+	private static final int MAX_LENGTH = Integer.MAX_VALUE;
+
+	private static final int DEFAULT_LENGTH = 1;
+
+	private static final String DEFAULT_FORMAT = "VARCHAR(%d)";
```

Comment:
nit: Is it a `DEFAULT_format`  or just a `FORMAT` can we change the `FORMAT`?

Reply:
It depends on the type. E.g. `ROW` and `ARRAY` could be represented in different ways. But I will correct it to `FORMAT` while merging.